### PR TITLE
Needinfo for bugs with high security keywords and low severity

### LINKED ISSUE #1336: Propose increasing severity for sec-high/sec-critical bugs that are lower than S2

### PR DESCRIPTION
**Pull request by suhaibmujahid**
This should resolve #1336.

I do not have access to security bugs. Thus I was not able to test the results. The following are needed:

- [x] Check the [query results](https://bugzilla.mozilla.org/buglist.cgi?include_fields=triage_owner&include_fields=assigned_to&include_fields=severity&include_fields=keywords&include_fields=id&include_fields=summary&include_fields=groups&include_fields=flags&resolution=---&f3=keywords&o3=anyexact&v3=sec-high&v3=sec-critical&f4=bug_severity&o4=anyexact&v4=S3&v4=S4&n15=1&f15=longdesc&o15=casesubstring&v15=could+you+consider+increasing+the+severity+of+this+security+bug%3F&f16=status_whiteboard&o16=notsubstring&v16=%5Bno-nag%5D&product=Core&product=DevTools&product=External+Software+Affecting+Firefox&product=Fenix&product=Firefox&product=Firefox+for+iOS&product=Firefox+Build+System&product=GeckoView&product=JSS&product=NSPR&product=NSS&product=Remote+Protocol&product=Testing&product=Toolkit&product=Web+Compatibility&product=WebExtensions).
- [x] Test the tool in the dry-run mode.

---

## Autonag wiki page

**Bugs with high security keywords which are set to low severity**

| **Purpose** | Identify bugs whose severity field is set to S3 or S4 and at the same time have sec-high or sec-critical keywords. |
| ----------- | :------------------------------------------------------------------------------------------------------------------ |
| **Action**  | Needinfo to propose increasing the severity.                                                                        |
| **Code**    | severity_high_security.py                                                                                           |

### CODE REVIEW

File: auto_nag/scripts/severity_high_security.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        bugid = str(bug["id"])
+        data[bugid] = {
+            "security_keyword": security_keyword,
+            "severity": bug["severity"],
+        }
+
+        return bug
+
+    def get_extra_for_needinfo_template(self):
+        return self.get_data()
```
Comment (marco-c) [inline auto_nag/scripts/severity_high_security.py:34]: This is going to return an empty dict, which will make `extra[bugid]["severity"]` in the needinfo template fail (as bugid is not in extra, given that extra is an empty dict).

In this script you don't actually need to implement this `get_extra_for_needinfo_template`, and you can simply replace `extra[bugid]["severity"]` with `severity` (and similarly for `extra[bugid]["security_keyword"]`.

File: auto_nag/scripts/severity_high_security.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+        params = {
+            "include_fields": fields,
+            "resolution": "---",
+            "f3": "keywords",
+            "o3": "anyexact",
+            "v3": HIGH_SECURITY_KEYWORDS,
+            "f4": "bug_severity",
+            "o4": "anyexact",
+            "v4": ["S3", "S4"],
```
Comment (marco-c) [inline auto_nag/scripts/severity_high_security.py:58]: Nit: maybe also include the old style severities such as `normal`, `minor`, etc.?

File: auto_nag/scripts/severity_high_security.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+from auto_nag import utils
+from auto_nag.bzcleaner import BzCleaner
+
+HIGH_SECURITY_KEYWORDS = ["sec-high", "sec-critical"]
+
+
+class SeverityHighSecurity(BzCleaner):
+    def description(self):
+        return "Bugs with high security keywords whose set to low severity"
```
Comment (marco-c) [inline auto_nag/scripts/severity_high_security.py:13]:
```suggestion
        return "Bugs with high security keywords which are set to low severity"
```

**Review by marco-c — CHANGES_REQUESTED** [review]
Need to fix the template application failure.

File: auto_nag/scripts/severity_high_security.py (hunk 1 above)
Comment (suhaibmujahid) [reply auto_nag/scripts/severity_high_security.py:34]: In the template rendering, only `bugid` and `extra` are passed:

https://github.com/suhaibmujahid/relman-auto-nag/blob/d23b911d8aa428dde533e7e6e38c9182e21896b0/auto_nag/bzcleaner.py#L440-L446

I will use the same workaround that I used in `severity_inconsistency.py`:

https://github.com/suhaibmujahid/relman-auto-nag/blob/08f0ce1d2193345706cff8ba253d01fe05929ed0/auto_nag/scripts/severity_inconsistency.py#L30-L35

Maybe a future refactoring should address this :)

Comment (suhaibmujahid) [reply auto_nag/scripts/severity_high_security.py]: Fixed in 7836438dc33ab217865b339d8f1f54295463b98d

File: auto_nag/scripts/severity_high_security.py (hunk 2 above)
Comment (suhaibmujahid) [reply auto_nag/scripts/severity_high_security.py]: Done in b25281f8427581e2d65495d6fb7fac1f72830c9c

File: auto_nag/scripts/severity_high_security.py (hunk 1 above)
Comment (marco-c) [reply auto_nag/scripts/severity_high_security.py]: What I suggested was actually working for me (I tested it locally)

Comment (suhaibmujahid) [reply auto_nag/scripts/severity_high_security.py]: Done in f65679d37c7128f47ff4db0389fc796ee5fbc5fa

**Review by marco-c — APPROVED** [review]

File: runauto_nag_daily.sh [hunk 4]
```diff
@@ -147,6 +147,9 @@ python -m auto_nag.scripts.severity_inconsistency --production
 # Needinfo for bugs with underestimated severity levels
 python -m auto_nag.scripts.severity_underestimated
 
+# Needinfo for bugs with high security keywords whose set to low severity
+python -m auto_nag.scripts.severity_underestimated
```
Comment (marco-c) [inline runauto_nag_daily.sh:151]: I just noticed that this should actually have been `severity_high_security`, and there's a missing `--production` (here and in the pre-existing `severity_underestimated` from above).

Comment (suhaibmujahid) [reply runauto_nag_daily.sh:151]: Nice catch, thank you. I fixed it in #1349.

**Pull request merged**